sqlx-sqlite/src/connection/mod.rs

```diff
@@ -112,6 +115,50 @@ impl SqliteConnection {
 
         Ok(LockedSqliteHandle { guard })
     }
+
+    /// Serializes the SQLite database to an in-memory buffer.
+    ///
+    /// This function allows you to serialize an SQLite database to an in-memory buffer.
+    /// The output is the same bytes as if the database were to be written to disk.
+    pub async fn serialize_to_buffer(&mut self) -> Result<SerializedDatabase, Error> {
+        let locked = self.lock_handle().await?;
+
+        unsafe {
+            let mut size: sqlite3_int64 = 0;
+            let data_ptr = sqlite3_serialize(
+                locked.guard.handle.as_ptr(),
+                std::ptr::null(),
+                &mut size as *mut _,
+                0,
+            );
+
+            Ok(SerializedDatabase::new(data_ptr, size as usize))
+        }
+    }
+    
+    /// Deserializes a byte slice into a read-only SQLite database.
+    ///
+    /// This function allows you to take a byte slice representing a serialized database,
+    /// and load it as a read-only SQLite database.
+    pub async fn deserialize_to_readonly_db(&mut self, data: &[u8]) -> Result<(), Error> {
+        let locked = self.lock_handle().await?;
+
+        unsafe {
+            let status = sqlite3_deserialize(
+                locked.guard.handle.as_ptr(),
+                std::ptr::null(),
+                data.as_ptr() as *mut u8,
+                data.len() as i64,
+                data.len() as i64,
+                SQLITE_DESERIALIZE_READONLY as u32,
+            );
+
+            match status {
+                SQLITE_OK => Ok(()),
+                _ => Err(SqliteError::new(locked.guard.handle.as_ptr()).into()),
+            }
+        }
+    }
```

Comment:
These functions could live in a standalone impl block in `serialized_database.rs` and then the constructor of `SerializedDatabase` can be completely private.